app/service/data_svc.py

```diff
@@ -391,6 +391,28 @@ def __init__(self, dao):
         """
         await self.dao.update(table, key, value, data)
 
+    async def update_agent_executor(self, agent_id, new_executors, previous_executors):
+        """
+        Update the agent's core executor
+        :param agent_id: string with the agent_id for the agent calling back
+        :param new_executors: list with incoming executors being reported by deployed agent
+        :param previous_executors: list of dict with previous executors and their preferred status
+        :return: None
+        """
+        old_executors = [d['executor'] for d in(sorted(previous_executors, key = lambda i: i['preferred'],
+                                                       reverse = True))]
+
+        for item in set(new_executors) - set(old_executors):
+            await self.dao.create('core_executor', dict(agent_id=agent_id, executor=item, preferred=0))
+
+        if old_executors[0] != new_executors[0]:
+            await self.dao.update('core_executor', 'agent_id', agent_id, data=dict(preferred=0))
+            await self.dao.update('core_executor', 'agent_id', str(agent_id) +
+                                  '" AND executor = "{}'.format(new_executors[0]), data=dict(preferred=1))
```

Comment:
let's keep all sql construction in the database class.

try adding a unique key on the executors table. then you can do a create here and the preference will be reflected properly.